{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create a manifest file\n",
    "\n",
    "The *manifest.txt* file created here is intended to be used with [gdc-client](https://gdc.cancer.gov/access-data/gdc-data-transfer-tool)\n",
    "\n",
    "Please refer to [gdc API](https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/) for reference about query done here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Author: Filippo_Valle\n",
      "\n",
      "Last updated: Mon Dec 07 2020\n",
      "\n",
      "Python implementation: CPython\n",
      "Python version       : 3.8.6\n",
      "IPython version      : 7.19.0\n",
      "\n",
      "pandas  : 1.1.4\n",
      "requests: 2.24.0\n",
      "json    : 2.0.9\n",
      "\n",
      "Compiler    : GCC 9.3.0\n",
      "OS          : Linux\n",
      "Release     : 5.4.39-linuxkit\n",
      "Machine     : x86_64\n",
      "Processor   : x86_64\n",
      "CPU cores   : 2\n",
      "Architecture: 64bit\n",
      "\n",
      "Git hash: c31e7538af2d650caf235a80152a3cf4a3c2e784\n",
      "\n",
      "Git repo: git@github.com:fvalle1/cancers\n",
      "\n",
      "Git branch: develop_first_round\n",
      "\n",
      "Watermark: 2.1.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -m  -u -n -p pandas,requests,json -a Filippo_Valle -g -r -b -w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use file endopint\n",
    "files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "# The 'fields' parameter is passed as a comma-separated string of single names.\n",
    "fields = [\n",
    "    \"file_name\",\n",
    "    #\"analysis.workflow_type\",\n",
    "    #\"experimental_strategy\",\n",
    "    \"cases.project.primary_site\",\n",
    "    \"cases.project.disease_type\",\n",
    "    \"cases.project.project_id\",\n",
    "    \"cases.submitter_id\",\n",
    "    \"cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"cases.demographic.vital_status\",\n",
    "    \"cases.demographic.days_to_birth\",\n",
    "    \"cases.demographic.days_to_death\",\n",
    "    \"cases.demographic.gender\",\n",
    "    \"cases.diagnoses.primary_diagnosis\",\n",
    "    \"cases.diagnoses.tumor_stage\",\n",
    "    \"cases.diagnoses.tumor_grade\",\n",
    "    \"cases.diagnoses.last_known_disease_status\",\n",
    "    \"cases.diagnoses.progression_or_recurrence\"\n",
    "    \"cases.diagnoses.days_to_last_follow_up\",\n",
    "    \"cases.diagnoses.age_at_diagnosis\",\n",
    "    \"cases.diagnoses.morphology\",\n",
    "    \"cases.diagnoses.tissue_or_organ_of_origin\",\n",
    "    \"cases.exposures.years_smoked\",\n",
    "    \"cases.exposures.cigarettes_per_day\",\n",
    "    \"cases.samples.longest_dimension\",\n",
    "    \"samples.portions.analytes.aliquots.submitter_id\",\n",
    "]\n",
    "\n",
    "fields = ','.join(fields)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter gene expression data obtained with *RNA-Seq HTSeq* in txt format either from **BRCA** or **LUAD** and **LUSC**\n",
    "\n",
    "Don't forget to select the correct filter for *cases.project.project_id*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tissue project filters\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"Gene Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [\"HTSeq - FPKM\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.project_id\",\n",
    "            \"value\": ['TCGA-LUAD', 'TCGA-LUSC'] #<-- uncomment this line for Lung\n",
    "            #\"value\": [\"TCGA-BRCA\"] #<-- uncomment this line for Breast\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "##remove manifest to have file fields\n",
    "params = {\n",
    "   \"return_type\": \"manifest\",\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The parameters are passed to 'json' rather than 'params' in this case\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "\n",
    "#print(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"manifest.txt\",\"w\") as manifest:\n",
    "    manifest.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Files\n",
    "\n",
    "Write a file with all metadata, useful in next analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"filters\": json.dumps(filters),\n",
    "    \"fields\": \"primary_site,cases.project.project_id,file_name,cases.demographic.vital_status,cases.demographic.gender,cases.diagnoses.age_at_diagnosis,cases.diagnoses.days_to_last_follow_up,cases.demographic.days_to_death,cases.demographic.days_to_birth,cases.submitter_id,samples.portions.analytes.aliquots.submitter_id,cases.diagnoses.last_known_disease_status,cases.diagnoses.tumor_stage,cases.exposures.years_smoked,cases.exposures.cigarettes_per_day,cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "with open(\"files.txt\",\"w\") as files:\n",
    "    files.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cases.0.demographic.days_to_birth</th>\n",
       "      <th>cases.0.demographic.days_to_death</th>\n",
       "      <th>cases.0.demographic.gender</th>\n",
       "      <th>cases.0.demographic.vital_status</th>\n",
       "      <th>cases.0.diagnoses.0.age_at_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.days_to_last_follow_up</th>\n",
       "      <th>cases.0.diagnoses.0.last_known_disease_status</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_stage</th>\n",
       "      <th>cases.0.exposures.0.cigarettes_per_day</th>\n",
       "      <th>cases.0.exposures.0.years_smoked</th>\n",
       "      <th>cases.0.project.project_id</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "      <th>cases.0.submitter_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>file_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2109.0</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage ia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-55-6980-11A-01R-1949-07</td>\n",
       "      <td>TCGA-55-6980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz</th>\n",
       "      <td>-26444.0</td>\n",
       "      <td>1288.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Dead</td>\n",
       "      <td>26444.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage ib</td>\n",
       "      <td>2.739726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-50-6590-01A-12R-1858-07</td>\n",
       "      <td>TCGA-50-6590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz</th>\n",
       "      <td>-26289.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Alive</td>\n",
       "      <td>26289.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage ia</td>\n",
       "      <td>2.191781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-69-8255-01A-11R-2287-07</td>\n",
       "      <td>TCGA-69-8255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  cases.0.demographic.days_to_birth  \\\n",
       "file_name                                                                             \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                                NaN   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                           -26444.0   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                           -26289.0   \n",
       "\n",
       "                                                  cases.0.demographic.days_to_death  \\\n",
       "file_name                                                                             \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                                NaN   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                             1288.0   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                                NaN   \n",
       "\n",
       "                                                 cases.0.demographic.gender  \\\n",
       "file_name                                                                     \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                       male   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                     female   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                       male   \n",
       "\n",
       "                                                 cases.0.demographic.vital_status  \\\n",
       "file_name                                                                           \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                            Alive   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                             Dead   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                            Alive   \n",
       "\n",
       "                                                  cases.0.diagnoses.0.age_at_diagnosis  \\\n",
       "file_name                                                                                \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                                   NaN   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                               26444.0   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                               26289.0   \n",
       "\n",
       "                                                  cases.0.diagnoses.0.days_to_last_follow_up  \\\n",
       "file_name                                                                                      \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                                      2109.0   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                                         NaN   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                                       129.0   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.last_known_disease_status  \\\n",
       "file_name                                                                                        \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                                  not reported   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                                  not reported   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                                  not reported   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.tumor_stage  \\\n",
       "file_name                                                                          \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                        stage ia   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                        stage ib   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                        stage ia   \n",
       "\n",
       "                                                  cases.0.exposures.0.cigarettes_per_day  \\\n",
       "file_name                                                                                  \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                                     NaN   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                                2.739726   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                                2.191781   \n",
       "\n",
       "                                                  cases.0.exposures.0.years_smoked  \\\n",
       "file_name                                                                            \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                               NaN   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                               NaN   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                               NaN   \n",
       "\n",
       "                                                 cases.0.project.project_id  \\\n",
       "file_name                                                                     \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                  TCGA-LUAD   \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                  TCGA-LUAD   \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                  TCGA-LUAD   \n",
       "\n",
       "                                                 cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \\\n",
       "file_name                                                                                                          \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz                       TCGA-55-6980-11A-01R-1949-07                \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz                       TCGA-50-6590-01A-12R-1858-07                \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz                       TCGA-69-8255-01A-11R-2287-07                \n",
       "\n",
       "                                                 cases.0.submitter_id  \n",
       "file_name                                                              \n",
       "e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.FPKM.txt.gz         TCGA-55-6980  \n",
       "96a02b1f-421c-40ee-8551-2b2f18709bcf.FPKM.txt.gz         TCGA-50-6590  \n",
       "a7ac2239-f60b-40a2-b931-5aa9b3adc8d8.FPKM.txt.gz         TCGA-69-8255  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_files = pd.read_csv(\"files.txt\", sep='\\t')\n",
    "df_files.drop(\"id\", axis=1, inplace=True)\n",
    "df_files.set_index(\"file_name\", inplace=True)\n",
    "df_files.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['cases.0.demographic.days_to_birth',\n",
       "       'cases.0.demographic.days_to_death', 'cases.0.demographic.gender',\n",
       "       'cases.0.demographic.vital_status',\n",
       "       'cases.0.diagnoses.0.age_at_diagnosis',\n",
       "       'cases.0.diagnoses.0.days_to_last_follow_up',\n",
       "       'cases.0.diagnoses.0.last_known_disease_status',\n",
       "       'cases.0.diagnoses.0.tumor_stage',\n",
       "       'cases.0.exposures.0.cigarettes_per_day',\n",
       "       'cases.0.exposures.0.years_smoked', 'cases.0.project.project_id',\n",
       "       'cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id',\n",
       "       'cases.0.submitter_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_files.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save file to dat\n",
    "df_files.to_csv(\"files_tcga.dat\", header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
